mat64/solve_test.go

```diff
@@ -0,0 +1,289 @@
+
+package mat64
+
+import (
+	"fmt"
+	"math/rand"
+	"os"
+
+	"gopkg.in/check.v1"
+)
+
+func (s *S) TestSolve(c *check.C) {
+	// Hand-coded cases.
+	for _, test := range []struct {
+		a         [][]float64
+		b         [][]float64
+		ans       [][]float64
+		shouldErr bool
+	}{
+		{
+			a:         [][]float64{{6}},
+			b:         [][]float64{{3}},
+			ans:       [][]float64{{0.5}},
+			shouldErr: false,
+		},
+		{
+			a: [][]float64{
+				{1, 0, 0},
+				{0, 1, 0},
+				{0, 0, 1},
+			},
+			b: [][]float64{
+				{3},
+				{2},
+				{1},
+			},
+			ans: [][]float64{
+				{3},
+				{2},
+				{1},
+			},
+			shouldErr: false,
+		},
+		{
+			a: [][]float64{
+				{0.8147, 0.9134, 0.5528},
+				{0.9058, 0.6324, 0.8723},
+				{0.1270, 0.0975, 0.7612},
+			},
+			b: [][]float64{
+				{0.278},
+				{0.547},
+				{0.958},
+			},
+			ans: [][]float64{
+				{-0.932687281002860},
+				{0.303963920182067},
+				{1.375216503507109},
+			},
+			shouldErr: false,
+		},
+		{
+			a: [][]float64{
+				{0.8147, 0.9134, 0.5528},
+				{0.9058, 0.6324, 0.8723},
+			},
+			b: [][]float64{
+				{0.278},
+				{0.547},
+			},
+			ans: [][]float64{
+				{0.25919787248965376},
+				{-0.25560256266441034},
+				{0.5432324059702451},
+			},
+			shouldErr: false,
+		},
+		{
+			a: [][]float64{
+				{0.8147, 0.9134, 0.9},
+				{0.9058, 0.6324, 0.9},
+				{0.1270, 0.0975, 0.1},
+				{1.6, 2.8, -3.5},
+			},
+			b: [][]float64{
+				{0.278},
+				{0.547},
+				{-0.958},
+				{1.452},
+			},
+			ans: [][]float64{
+				{0.820970340787782},
+				{-0.218604626527306},
+				{-0.212938815234215},
+			},
+			shouldErr: false,
+		},
+		{
+			a: [][]float64{
+				{0.8147, 0.9134, 0.231, -1.65},
+				{0.9058, 0.6324, 0.9, 0.72},
+				{0.1270, 0.0975, 0.1, 1.723},
+				{1.6, 2.8, -3.5, 0.987},
+				{7.231, 9.154, 1.823, 0.9},
+			},
+			b: [][]float64{
+				{0.278, 8.635},
+				{0.547, 9.125},
+				{-0.958, -0.762},
+				{1.452, 1.444},
+				{1.999, -7.234},
+			},
+			ans: [][]float64{
+				{1.863006789511373, 44.467887791812750},
+				{-1.127270935407224, -34.073794226035126},
+				{-0.527926457947330, -8.032133759788573},
+				{-0.248621916204897, -2.366366415805275},
+			},
+			shouldErr: false,
+		},
+		{
+			a: [][]float64{
+				{0, 0},
+				{0, 0},
+			},
+			b: [][]float64{
+				{3},
+				{2},
+			},
+			ans:       nil,
+			shouldErr: true,
+		},
+	} {
+		a := NewDense(len(test.a), len(test.a[0]), nil)
+		for i := 0; i < len(test.a); i++ {
+			for j := 0; j < len(test.a[0]); j++ {
+				a.Set(i, j, test.a[i][j])
+			}
+		}
+		b := NewDense(len(test.b), len(test.b[0]), nil)
+		for i := 0; i < len(test.b); i++ {
+			for j := 0; j < len(test.b[0]); j++ {
+				b.Set(i, j, test.b[i][j])
+			}
+		}
+		var ans *Dense
+		if test.ans != nil {
+			ans = NewDense(len(test.ans), len(test.ans[0]), nil)
+			for i := 0; i < len(test.ans); i++ {
+				for j := 0; j < len(test.ans[0]); j++ {
+					ans.Set(i, j, test.ans[i][j])
+				}
+			}
+		}
+
+		var x Dense
+		err := x.Solve(a, b)
+		if err != nil {
```

Comment:
No, it's not the same. If err != nil, then we don't want to compare x with the answer. We want to continue if err != nil regardless of test.shouldErr